File: ts/src/parkingBoy/mostAvailableParkingBoyRule.ts
/** @format */

import {ParkingBoyRule} from './parkingBoyRule'
import {ParkingLot} from '../parkingLot/parkingLot'

export class MostAvailableParkingBoyRule implements ParkingBoyRule {
    public findOneValidParkingLot(parkingLots: ParkingLot[]) {
        return (
            parkingLots
                .filter(({hasAvailableSpaces}) => hasAvailableSpaces)
                .sort(({availableSpaces}) => -availableSpaces)[0] || null
        )
    }
}

File: kt/src/main/kotlin/com/github/scglwsj/parking/boy/MostAvailableParkingBoyRule.kt
package com.github.scglwsj.parking.boy

import com.github.scglwsj.parking.lot.ParkingLot

class MostAvailableParkingBoyRule : ParkingBoyRule {
    override fun findOneValidParkingLot(parkingLots: List<ParkingLot>): ParkingLot? =
        parkingLots.sortedByDescending { it.availableSpaces }.firstOrNull { it.hasAvailableSpaces }
}
